realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java

```diff
@@ -301,7 +349,12 @@ public void execute(Realm realm) {
             @Override
             public void onSuccess() {
                 assertEquals(1, realm.where(AllTypes.class).count());
-                assertEquals(1, results.size());
+                // We cannot guarantee the async results get delivered from OS.
```

Comment:
I think I missed that the result was coming from an async query, which I agree complicates matters quite a lot. I read it as the thread hasn't been updated if you did a synchronous query, which is obviously wrong.

I'm still a bit concerned about this as right now I think we postpone transaction callbacks until all queries have been re-run?

Regarding your case 2) I'm not sure I fully understand it. You mean touching any element in it will make the query run on the IU thread? (which brings be back to my first point).